src/server.py

```diff
@@ -255,16 +255,41 @@ def _add_keyframe_pair(self, control, time_start_ns, value_start,
     def _add_keyframe(self, control, time_ns, value):
         return control.set(time_ns, value)
 
-    def _add_fade_in(self, time_ms_end, volume_end):
-        self._add_keyframe_pair(self._fade_control, 0, 0,
-                                time_ms_end * Gst.MSECOND, volume_end, False)
-
-    def _add_fade_out(self, element, time_ms):
-        current_volume = element.props.volume
+    def _add_fade_in(self):
+        if not self.loop or self.loop and self.fade_in == 0:
+            return
+        self.logger.debug("Fading in.")
+        try:
+            current_time = self.get_current_position()
+        except ValueError:
+            # This is the first call to play the sound and it should fade in
+            # from the time 0.
+            self.logger.info("Assume first PlaySound call. Current time=0.")
+            current_time = 0
+        current_volume = self.pipeline.get_by_name("volume").props.volume
+        end_time = current_time + self.fade_in * Gst.MSECOND
+        self._add_keyframe_pair(self._fade_control,
+                                current_time, current_volume,
+                                end_time, self.volume, False)
+
+    def _add_fade_out(self):
+        # This method may raise a ValueError usually if the pipeline is in
```

Comment:
Shouldn't we catch this exception if it happens, and stop the in-progress state change, like we catch the ValueError in `_add_fade_in`? Or is it precisely because we can't do anything about it that we don't catch it? Maybe catch it and log it in that case?

Reply:
In `_add_fade_in`, what I do is to assume that if there is an error, then it must be the **first** fade in.  Because in the first fade in, the pipeline will be in READY state so the current position (that actually is 0, before the pipeline gets played) cannot be queried.

I don't like the idea of aborting the in-progress state change. Because I guess we would have to register that pending state change to resume it later. Also, It will make things more complicated since the state change happens in another thread.

>  Maybe catch it and log it in that case?
That is possible. But... do you mean to log the current state or the pending state change? Both are possible to log.

Reply:
I am logging the current state. If you have other opinion, let me know.

Reply:
My comment was more about not letting the method throw an exception if there wasn't anything that the caller could do about it. For example, my understanding is that if you call PlaySound and then StopSound immediately, that exception may be thrown, or it may not, depending on how fast the calls arrive over the bus. So, should that be an invalid thing for a caller to do, or not? I think it should be valid, because it's not the caller's problem that the state is still NULL or READY, so we should catch that exception and do what the caller intends, which is stop the sound.

I might be overlooking it, but I don't see the change you mentioned in your most recent comment...

Reply:
I don't agree. Because the _add_fade_out method could be used to also pause the sound with fade out (bg sounds rule). What I would agree with would be to call `self.release()` just after https://github.com/endlessm/hack-sound-server/blob/T25659/src/server.py#L129

Reply:
Let me know what you think

Reply:
What will that do?

If you think it's better to do something else than stop the sound immediately, then I could be convinced, as long as we avoid these two things:
- API calls should not result in exceptions if the caller can't do anything about them.
- The intended effect of an API call should not be ignored. (e.g. if StopSound is called and the sound never stops)

Reply:
Calling `self.release` just after https://github.com/endlessm/hack-sound-server/blob/T25659/src/server.py#L129 would immediately stop the sound if any exception occurs.

Reply:
> API calls should not result in exceptions if the caller can't do anything about them.

Adding self.release() solves that, since if the _add_fade_out function fails, then it will stop the sound immediately.

> API calls should not result in exceptions if the caller can't do anything about them.

Because even if _add_fade_out raises an exception the pipeline is released/stopped, the StopSound call is not being ignored.

Reply:
OK, I see. The exception is caught elsewhere. I was assuming that you would change this method to not throw the exception anymore, because also the caller _of this method_ can't predict or control whether such an exception would be thrown. However, I'm okay with compromising on that for now, since I think this commit is at least an improvement on what we have now.